src/std/actor-v18/proto.ss

```diff
@@ -0,0 +1,97 @@
+;;; -*- Gerbil -*-
+;;; © vyzo
+;;; actor protocol messages
+(import :gerbil/gambit/threads
+        :std/sugar
+        ./message)
+(export #t)
+
+(defmessage !ok (value))
+(defmessage !error (message))
+
+(defrules with-result ()
+  ((_ expr fail!)
+   (match expr
+     ((!ok value) value)
+     ((!error what) (fail! what))))
+  ((_ expr)
+   (with-result expr error)))
+
+(defrules defcall-actor ()
+  ((_ (proc arg ...) expr error: error-msg error-irritant ...)
+   (def (proc arg ...)
+     (with-result expr
+       (lambda (what) (raise-actor-error 'proc error-msg error-irritant ... what)))))
+  ((_ (proc arg ...) expr)
+   (defcall-actor (proc arg ...) expr error: "actor error")))
+
+(defmessage !shutdown ())
+(defmessage !actor-dead (thread))
+(defmessage !tick (id seqno))
+(defmessage !ping ())
+
+;; reaction macro for ping
+(defrule (@ping)
+  ((!ping) (--> (!ok 'OK))))
+
+;; reaction macro for unexpected messages
+(defrule (@unexpected logf)
+  (unexpected
+   (logf "unexpected message from ~a: ~a" @source @message)
+   (-->? (!error "unexpected message"))))
+
+;; reaction macro for shutdown
+(defrule (@shutdown exit ...)
+  ((!shutdown)
+   (-->? (!ok (void)))
+   exit ...))
+
+;; package private
```

Comment:
There again, some documentation on expected types, etc., would be nice.

Reply:
see reference, I don't see the need to clutter the code at this point. Feel free to do a later pass and add comments as you are reading the code (and ask questions).